docs/user-guide/setting-up-ci-cd.rst

```diff
@@ -0,0 +1,212 @@
+
+.. code-block:: bash
+
+   rasa data validate --fail-on-warnings --max-history 5
+
+Conflicts between stories will prevent Rasa from learning the correct pattern.
+You can run story validation by passing the ``--max-history`` flag to ``rasa
+data validate``, either in a seperate check or as part of the data validation
+check. ``--max-history`` should be set to the value of ``max_history`` in your
+config for your memoization policy (which defaults to ``5``).
+
+Train a Model
+#############
+
+.. code-block:: bash
+
+   rasa train
+
+Training a model verifies that your NLU pipeline and policy configurations are
+valid and trainable, and it provides a model to use for test conversations. Training a model is also :ref:`part of the continuous deployment
+process <uploading-a-model>`, as you'll need a model to upload to your server. 
+
+Test the Assistant
+##################
+
+Testing your trained model on :ref:`test conversations
+<end-to-end-testing>` is the best way to have confidence in how your assistant
+will act in certain situations. These stories, written in a modified story
+format, allow you to provide entire conversations and test that, given this
+user input, your model will behave in the expected manner. This is especially
+important as you start introducing more complicated stories from user
+conversations. End-to-end testing is only as thorough and accurate as the test
+cases you write, so you should always update your test conversations
+whenever you make changes to your training stories.
+
+.. code-block:: bash
+
+   rasa test --stories tests/test_stories.md --fail-on-prediction-errors
+
+To ensure the test will fail if any test conversation fails, add 
+the ``--fail-on-prediction-errors`` flag:
+
+Note: End-to-end testing does **not** execute your action code. You will need to
+:ref:`test your action code <testing-action-code>` in a seperate step.
+
+Compare NLU Performance
+#######################
+
+If you've made significant changes to your NLU training data (such as adding or
+splitting intents, or adding/changing a lot of examples), you should run a
+:ref:`full NLU evaluation <nlu-evaluation>`. You'll want to compare
+the performance of the NLU model without your changes to an NLU model with your
+changes. 
+
+You can do this by running NLU testing in cross-validation mode:
+
+.. code-block:: bash
+
+   rasa test nlu --cross-validation
+
+or by training a model on a training set and testing it on a test set. If you use the train-test
+set approach, it is best to :ref:`shuffle and split your data <train-test-split>` using ``rasa data split`` as part of this CI step, as
+opposed to using a static NLU test set, which can easily become outdated. 
+
+Since NLU comparison can be a fairly resource intensive test, you can set this
+test to run only when a certain tag (e.g. "NLU testing required") is present,
+or only when changes to NLU data or the NLU pipeline were made. 
+
```

Comment:
I might add that a good practice is to have the CI stage produce an artifact that to be reviewed before the changes are merged in (e.g. https://gist.github.com/amn41/de555c93913a01fbd56df2e2d211862c)